File: 0329. Longest Increasing Path in a Matrix/Solution0329.py
from typing import List


class Solution0329:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        dp = [[1 for i in range(n)] for i in range(m)]
        longest = 1
        for i in range(m):
            for j in range(n):
                longest = max(longest, self.dfs(i, j, matrix, dp))
        return longest

        
    def dfs(self, r: int, c: int, matrix: List[List[int]], dp: List[List[int]]) -> int: 
        m, n = len(matrix), len(matrix[0])
        if dp[r][c] != 1:
            return dp[r][c]
        directions = [(1,0), (-1,0), (0,1), (0,-1)]
        for dr, dc in directions:
            nextr = r + dr
            nextc = c + dc
            if nextr < 0 or nextr >= m or nextc < 0 or nextc >= n or matrix[nextr][nextc] <= matrix[r][c]:
                continue
            dp[r][c] = max(dp[r][c], self.dfs(nextr, nextc, matrix, dp)+1)

        return dp[r][c]
        
        

File: 0329. Longest Increasing Path in a Matrix/Solution0329.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution0329 {
public:
    int DIRS[4][2] = {{0,1},{1,0},{0,-1},{-1,0}};    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int row = matrix.size(), col = matrix[0].size();
        int longest = 0;        
        vector<vector<int>> dp(row, vector<int>(col));
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                longest = max(longest, dfs(i, j, dp, matrix)); //
            }
        }    
        return longest;
    }
    
    int dfs(int r, int c, vector<vector<int>>& dp, vector<vector<int>>& matrix) {
        int row = matrix.size(), col = matrix[0].size();
        if (dp[r][c] != 0) return dp[r][c];
        int res = 0;
        for (auto dir : DIRS) {
            int nr = r + dir[0], nc = c + dir[1];
            if (nr >= 0 && nr < row && nc >= 0 && nc < col && matrix[nr][nc] < matrix[r][c]) {
                res = max(res, dfs(nr, nc, dp, matrix));
            }
        }
        dp[r][c] = res + 1;
        return dp[r][c];
    }
};